File: src/webhook.rs
use axum::{
    Json, Router,
    extract::{Request, State},
    http::{HeaderMap, StatusCode},
    response::{IntoResponse, Response},
    routing::{get, post},
};
use serde_json::{Value, json};

use monkey_core::config::Settings;
use monkey_core::db::Store;
use monkey_core::hmac_auth::verify_github_signature;

#[derive(Clone)]
pub struct WebhookState {
    pub settings: Settings,
    pub store: Store,
}

pub fn app(state: WebhookState) -> Router {
    Router::new()
        .route("/healthz", get(healthz))
        .route("/webhook/github", post(github_webhook))
        .with_state(state)
}

async fn healthz() -> Json<Value> {
    Json(json!({ "ok": true }))
}

async fn github_webhook(
    State(state): State<WebhookState>,
    headers: HeaderMap,
    req: Request,
) -> Response {
    let body = match axum::body::to_bytes(req.into_body(), 10 * 1024 * 1024).await {
        Ok(b) => b,
        Err(_) => {
            return (
                StatusCode::BAD_REQUEST,
                Json(json!({ "detail": "failed to read body" })),
            )
                .into_response();
        }
    };

    let signature = headers
        .get("x-hub-signature-256")
        .and_then(|v| v.to_str().ok());

    if verify_github_signature(&state.settings.github_webhook_secret, &body, signature).is_err() {
        return (
            StatusCode::UNAUTHORIZED,
            Json(json!({ "detail": "bad signature" })),
        )
            .into_response();
    }

    let delivery_id = match headers
        .get("x-github-delivery")
        .and_then(|v| v.to_str().ok())
    {
        Some(d) if !d.is_empty() => d,
        _ => {
            return (
                StatusCode::BAD_REQUEST,
                Json(json!({ "detail": "missing x-github-delivery" })),
            )
                .into_response();
        }
    };

    let event_type = headers
        .get("x-github-event")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("");

    let payload: Value = match serde_json::from_slice(&body) {
        Ok(v) => v,
        Err(_) => {
            return (
                StatusCode::BAD_REQUEST,
                Json(json!({ "detail": "invalid json" })),
            )
                .into_response();
        }
    };

    let (owner, repo, number) = match parse_target(event_type, &payload) {
        (Some(o), Some(r), Some(n)) => (o, r, n),
        _ => {
            return (
                StatusCode::OK,
                Json(json!({ "ok": true, "skipped": "not an actionable event" })),
            )
                .into_response();
        }
    };

    let target_repo = format!("{}/{}", owner, repo);
    if !state
        .settings
        .allowlist()
        .iter()
        .any(|allowed| allowed == &target_repo)
    {
        return (
            StatusCode::OK,
            Json(json!({ "ok": true, "skipped": "repo not in allowlist" })),
        )
            .into_response();
    }

    let sender = payload
        .get("sender")
        .and_then(|sender| sender.get("login"))
        .and_then(|login| login.as_str())
        .unwrap_or("");
    let sender_type = payload
        .get("sender")
        .and_then(|sender| sender.get("type"))
        .and_then(|sender_type| sender_type.as_str())
        .unwrap_or("");
    if (!state.settings.bot_login.is_empty()
        && sender.eq_ignore_ascii_case(&state.settings.bot_login))
        || sender.ends_with("[bot]")
        || sender_type.eq_ignore_ascii_case("bot")
    {
        return (
            StatusCode::OK,
            Json(json!({ "ok": true, "skipped": "bot-authored event" })),
        )
            .into_response();
    }

    let body_str = String::from_utf8_lossy(&body);
    let is_new =
        match state
            .store
            .enqueue(delivery_id, event_type, &owner, &repo, number, &body_str)
        {
            Ok(inserted) => inserted,
            Err(e) => {
                return (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    Json(json!({ "detail": format!("database error: {}", e) })),
                )
                    .into_response();
            }
        };

    (StatusCode::OK, Json(json!({ "ok": true, "new": is_new }))).into_response()
}

pub fn parse_target(
    event_type: &str,
    payload: &Value,
) -> (Option<String>, Option<String>, Option<i64>) {
    if !matches!(
        event_type,
        "issues" | "pull_request" | "issue_comment" | "pull_request_review"
    ) {
        return (None, None, None);
    }

    let repo_obj = match payload.get("repository") {
        Some(r) => r,
        None => return (None, None, None),
    };
    let owner = repo_obj
        .get("owner")
        .and_then(|o| o.get("login").or_else(|| o.get("name")))
        .and_then(|v| v.as_str())
        .map(|s| s.to_string());

    let repo = repo_obj
        .get("name")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string());

    let number = payload
        .get("issue")
        .or_else(|| payload.get("pull_request"))
        .or_else(|| payload.get("review"))
        .and_then(|item| item.get("number"))
        .and_then(|n| n.as_i64());

    (owner, repo, number)
}

File: legacy/monkey/webhook.py
"""GitHub webhook receiver.

Verifies HMAC, dedups on X-GitHub-Delivery, and enqueues the event for the
worker. This endpoint is the ONLY thing exposed to the internet.
"""

from __future__ import annotations

import json
from typing import Annotated

from contextlib import asynccontextmanager

from fastapi import FastAPI, Header, HTTPException, Request
from fastapi.responses import JSONResponse

from .config import get_settings
from .db import Store
from .hmac import BadSignature, verify_github_signature


@asynccontextmanager
async def lifespan(app: FastAPI):
    get_settings()  # raises if required config missing
    yield


app = FastAPI(title="monkey", lifespan=lifespan)


def _env_store() -> Store:
    # lite lifecycle: one Store per process, created lazily.
    global _STORE
    if _STORE is None:
        _STORE = Store("/data/monkey.db")
    return _STORE


_STORE: Store | None = None
@app.get("/healthz")
def healthz() -> dict:
    return {"ok": True}


@app.post("/webhook/github")
async def github_webhook(request: Request):
    settings = get_settings()
    store = _env_store()

    body = await request.body()
    signature = request.headers.get("x-hub-signature-256")
    delivery_id = request.headers.get("x-github-delivery")
    event_type = request.headers.get("x-github-event", "")

    try:
        verify_github_signature(settings.github_webhook_secret, body, signature)
    except BadSignature:
        # 401, never 5xx, so GitHub stops retrying a bad secret.
        return JSONResponse(status_code=401, content={"detail": "bad signature"})

    if not delivery_id:
        raise HTTPException(status_code=400, detail="missing x-github-delivery")

    try:
        payload = json.loads(body)
    except json.JSONDecodeError:
        raise HTTPException(status_code=400, detail="invalid json")

    owner, repo, number = _parse_target(event_type, payload)
    if owner is None or repo is None or number is None:
        # Non-issue/pr/runnable event (e.g. ping) - acknowledge, no work.
        return {"ok": True, "skipped": "not an actionable event"}

    # Allowlist check.
    if f"{owner}/{repo}" not in settings.allowlist:
        return {"ok": True, "skipped": "repo not in allowlist"}

    # Ignore events authored by the bot itself or other bots to avoid feedback loops.
    sender = (payload.get("sender") or {}).get("login", "")
    if settings.bot_login and sender.lower() == settings.bot_login.lower():
        return {"ok": True, "skipped": "bot-authored event"}
    if sender.endswith("[bot]"):
        return {"ok": True, "skipped": "bot-authored event"}

    is_new = store.enqueue(
        delivery_id, event_type, owner, repo, number, body.decode("utf-8", "replace")
    )
    return {"ok": True, "new": is_new}


def _parse_target(event_type: str, payload: dict) -> tuple[str | None, str | None, int | None]:
    """Extract (owner, repo, issue/pr number) for the events we care about."""
    repo = payload.get("repository") or {}
    owner = (repo.get("owner") or {}).get("login") or (repo.get("owner") or {}).get("name")
    repo_name = repo.get("name")

    def _num():
        n = payload.get("issue") or payload.get("pull_request") or payload.get("review") or {}
        return n.get("number")

    if event_type in ("issues", "pull_request", "issue_comment", "pull_request_review"):
        return owner, repo_name, _num()
    return None, None, None
